### PR TITLE
Improve AnalyticsFactory to handle all types of Request

### PR DESCRIPTION
**Pull request by rvitaliy**

### CODE REVIEW

**Comment by rvitaliy** [comment]
@foaly-nr1 yes, of course.

- [major] improve handle of `User-Agent` to manage different `$request->header->get()` results `null|string|string[]`
- [major] call `setIpOverride` only when we have a value !== null
- [major] call `setClientId ` only when we have a value !== null
- [minor] fixed phpDoc with right type hints
- [minor] improve test to be more strict using `assertSame` instead of `assertEquals`

File: Service/AnalyticsFactory.php [hunk 1]
```diff
@@ -83,32 +84,106 @@ public function createAnalytics()
...
+
+    /**
+     * @param Analytics $analytics
+     * @param Request   $request
+     */
+    private function handleClientIp(Analytics $analytics, Request $request)
+    {
+        $clientIp = $request->getClientIp();
+
+        if (null === $clientIp) {
```
Comment (foaly-nr1) [inline Service/AnalyticsFactory.php]: Again, is this a real-life scenario? There don't seem to be tests.

**Review by jonny-no1 — COMMENTED** [review]
Thanks for the clarification! I can see the additional cases you're covering in `handleUserAgent()` and `handleClientId()` but have some questions around some of the other behaviour you introduced.

File: Service/AnalyticsFactory.php [hunk 2]
```diff
@@ -83,32 +84,106 @@ public function createAnalytics()
...
     {
-        return explode('.', substr($cookie, 2), 3);
+        $parsedCookie = explode('.', substr($cookie, 2), 3);
+        $missingElements = 3 - count($parsedCookie);
+
+        if ($missingElements <= 0) {
+            return $parsedCookie;
+        }
+
+        return array_merge(
```
[the remark this answers is not in the document]
Comment (rvitaliy) [reply Service/AnalyticsFactory.php]: IMHO: it is.
Nothing prevents me from putting in the cookie the value I want. With this checks we prevent what our code generate an error and we garanty always to return an array with at least 3 elements containing strings.

File: Service/AnalyticsFactory.php [hunk 3]
```diff
@@ -83,32 +84,106 @@ public function createAnalytics()
...
-                $cookie = $this->parseCookie($ga);
-                $analytics->setClientId(array_pop($cookie));
-            }
+        if (null === $request) {
+            return $analytics;
         }
 
+        $this->handleClientIp($analytics, $request);
+        $this->handleUserAgent($analytics, $request);
+        $this->handleClientId($analytics, $request);
```
[the remark this answers is not in the document]
Comment (rvitaliy) [reply Service/AnalyticsFactory.php]: we call 2 different functions `handleClient`**Ip** `handleClient`**Id**

File: Service/AnalyticsFactory.php (hunk 1 above)
Comment (rvitaliy) [reply Service/AnalyticsFactory.php]: the signature of method `Request::getClientIp` says:
```
    /**
     * Returns the client IP address.
     *
     * This method can read the client IP address from the "X-Forwarded-For" header
     * when trusted proxies were set via "setTrustedProxies()". The "X-Forwarded-For"
     * header value is a comma+space separated list of IP addresses, the left-most
     * being the original client, and each successive proxy that passed the request
     * adding the IP address where it received the request from.
     *
     * @return string|null The client IP address
     *
     * @see getClientIps()
     * @see http://en.wikipedia.org/wiki/X-Forwarded-For
     */
```

IMHO: it's possible to have a nullable value.

File: Service/AnalyticsFactory.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (foaly-nr1) [reply Service/AnalyticsFactory.php]: I agree that we should check cookie values and ensure that it's in the expected format. But I don't think we should try to fill in gaps. So why don't we change line 113-115 to
```
if (count($parsedCookie) !== 3) {
    return null;
}
```

File: Service/AnalyticsFactory.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (foaly-nr1) [reply Service/AnalyticsFactory.php]: Sorry, my bad! Can you write tests for those two new methods please?

[the remark this answers is not in the document]
Comment (rvitaliy) [reply Service/AnalyticsFactory.php]: all this handler are private methods. and coverd by tests:
- `testParseCookieWithEmptyString`
- `testGaCookie`
- `testGaCookieWithEmptyString`
- `testGetIpOverrideIsNull`
- `testGetIpOverride`
- `testUserAgentOverride`

**Pull request closed without merging**